substrate/client/network/sync/src/pending_responses.rs

```diff
@@ -0,0 +1,192 @@
+}
+
+impl<B: BlockT> PendingResponses<B> {
+	pub fn new() -> Self {
+		Self {
+			pending_responses: HashMap::new(),
+			finished_responses: VecDeque::new(),
+			finished_peer_ids: HashSet::new(),
+			waker: None,
+		}
+	}
+
+	pub fn insert(
+		&mut self,
+		peer_id: PeerId,
+		request: PeerRequest<B>,
+		response_future: ResponseFuture,
+	) {
+		let request_type = request.get_type();
+
+		if let Some(_) = self.pending_responses.insert(
+			peer_id,
+			async move { ResponseEvent { peer_id, request, response: response_future.await } }
+				.boxed(),
+		) {
+			error!(
+				target: LOG_TARGET,
+				"Discarded pending response from peer {peer_id}, {request_type:?}.",
+			);
+			debug_assert!(false);
+		}
+
+		if self.finished_peer_ids.remove(&peer_id) {
+			// This must not happen unless there is a logic error in the code,
+			// so performance penalty is not an issue.
+			self.finished_responses.retain(|event| event.peer_id != peer_id);
+
+			error!(
+				target: LOG_TARGET,
+				"Discarded finished response from peer {peer_id}, {request_type:?}.",
+			);
+			debug_assert!(false);
+		}
+
+		if let Some(waker) = self.waker.take() {
+			waker.wake();
+		}
+	}
+
+	pub fn remove(&mut self, peer_id: &PeerId) -> bool {
+		match self.pending_responses.remove(peer_id) {
+			Some(_) => true,
+			None => {
+				// then we must look in finished responses
+				if self.finished_peer_ids.remove(peer_id) {
+					// We do a linear search here. With 10 ms roud-trip network latency and 100
+					// active peers this leads to a penalty of up to 10000 `PeerId` comparisons per
+					// second. Moreover, we remove an element from a `VecDeque`, what leads to a
+					// penalty proportional to the poll iteration duration (and hence the number of
+					// elements in a `VecDeque`).
+					// TODO: probably optimize this.
+
+					let mut remove_index = None;
+
+					for (i, event) in self.finished_responses.iter().enumerate() {
+						if &event.peer_id == peer_id {
+							remove_index = Some(i);
+							break
+						}
+					}
+
+					if let Some(index) = remove_index {
+						self.finished_responses.remove(index);
+						true
+					} else {
+						error!(
+							target: LOG_TARGET,
+							"Logic error: {peer_id} is in `finished_peer_ids`, but not in `finished_responses`.",
+						);
+						debug_assert!(false);
+						false
+					}
+				} else {
+					false
+				}
+			},
+		}
+	}
+
+	pub fn len(&self) -> usize {
+		self.pending_responses.len() + self.finished_responses.len()
+	}
+}
+
+impl<B: BlockT> Unpin for PendingResponses<B> {}
+
+impl<B: BlockT> Stream for PendingResponses<B> {
+	type Item = ResponseEvent<B>;
+
+	fn poll_next(
+		mut self: std::pin::Pin<&mut Self>,
+		cx: &mut Context<'_>,
+	) -> Poll<Option<Self::Item>> {
+		let ready_responses = self
+			.pending_responses
+			.values_mut()
+			.filter_map(|future| match future.poll_unpin(cx) {
+				Poll::Pending => None,
+				Poll::Ready(event) => Some(event),
+			})
+			.collect::<Vec<_>>();
```

Comment:
Why is it polling all of them in one go? Why not use a loop and the first one that is finished is returned to user? It looks like you could simplify the code quite a bit this way

Reply:
I tried to make it fair in terms of which pending responses are polled. If we just polled them in a loop and return the one that is finished, the responses in the beginning of the hash table would be polled much frequently than the responses at the end of the table. This could lead to starvation of responses that ended up at the end of the table.

Reply:
What about using [`StreamMap`](https://docs.rs/tokio-stream/latest/tokio_stream/struct.StreamMap.html) for storing pending responses? That would give us insert/remove capabilities and it implements `Stream` for us out of the box. It also randomizes the start index so all futures would be polled equally.

Reply:
Should be done now. I'm not sure though if it worth keeping `PendingResponses` as a separate struct: the only real help from it now is code de-duplication when reporting overwritten responses.

Reply:
As we basically do `response_future.into_stream()`, the contained stream yields only one item by design. So, i think there is no need for [`StreamNotifyClose`](https://docs.rs/tokio-stream/latest/tokio_stream/struct.StreamNotifyClose.html).